docs: move template ref vars doc to concepts section clarify, add scope section, and update headers (#31195)
angular/angular: aio/content/guide/template-syntax.md

<code-example path="template-reference-variables/src/app/app.component.html" region="ref-var" header="src/app/app.component.html" linenums="false">
</code-example>

Here, a `<button>` further down the template refers to the `phone` variable.

<code-example path="template-reference-variables/src/app/app.component.html" region="ref-phone" header="src/app/app.component.html" linenums="false">
</code-example>

<div class="alert is-helpful">

> You can refer to a template reference variable anywhere

Reviewer: Thanks for working on this! I would perhaps change a few things, though it's definitely already an improvement :-) 

> You can refer to a template reference variable anywhere directly within its surrounding template. Note, however, that structural directives such as `*ngIf` and `*ngFor`, or `<ng-template>` act as a template boundary outside of which the template reference variable cannot be accessed.

(With the links and formatting you added).

It's also an interesting fact that inheritance into templates within the same template definition do actually work, the example below shows that I mean. I think this is the case because the template context will be inherited. Nesting components is, one would think, the "same" – but will *not* work because the context is not inherited in this case. I have to admit that I'm not sure if there's a good distinction in the terminology of "templates" that makes this easier to explain.

```
<!-- This works -->
<input #ref type="text" />
<ng-container *ngIf="true">{{ ref.value }}</ng-container>

<!-- But this does not -->
<ng-container *ngIf="true">
  <input #ref type="text" />
</ng-container>

{{ ref.value }}
```

Perhaps we want to add these two examples to the docs here as well? It might be easier to understand than in words.

---

BTW: I have had to explain *why* this limitation exists to a few people, and I find that `*ngFor` is a much better example to do so than `*ngIf`:

```
<ng-container *ngFor="let i of [1,2]">
  <input #ref type="text" [value]="i" />
</ng-container>

<!-- Should this display "1" or "2"? -->
{{ ref.value }}
```

It shows the basic problem of accessing the variable from "outside" the template: there is no guarantee that the template is ever instantiated and, even worse, that it's instantiated only once, so accessing it simply doesn't make sense. For `*ngIf` it sort of would, but Angular doesn't treat `*ngIf` "special", so the framework has no way of knowing that the template will only be instantiated once.

I felt like maybe it's good to add this example in this comment here. I don't necessarily think we should add it to the docs, though.

Author: Wow, this is super helpful. Thank you, @Airblader. I've edited the wording as you suggested.

For this part:

```
<!-- This works -->
<input #ref type="text" />
<ng-container *ngIf="true">{{ ref.value }}</ng-container>

<!-- But this does not -->
<ng-container *ngIf="true">
  <input #ref type="text" />
</ng-container>

{{ ref.value }}
```

Could you help me with a StackBlitz for it? I tried but get lost pretty fast. I'd like to add a working example so that the code snippet isn't inline.

Reviewer: @kapunahelewong I've put a quick example together here: https://stackblitz.com/edit/angular-sjrq5h It's not the most beautifully designed StackBlitz ever, though. :-) Feel free to adapt it more!

Author: This is so wonderful, @Airblader! Thank you!! So that I understand, for this first part:
```
<p>Accessing a template reference variable from an inner template 
works as the context is inherited....</p>

<div class="example">
  <input #ref1 type="text" [(ngModel)]="firstExample" />
  <span *ngIf="true">Value: {{ ref1.value }}</span>
</div>
```

^^This works because we access `ref1` within the template created by the `*ngIf` (that is, we're still inside the `<span>`)?

```
<p>However, accessing it from outside the parent template will not work....</p>

<div class="example">
  <input *ngIf="true" #ref2 type="text" [(ngModel)]="secondExample" />
  <span>Value: {{ ref2?.value }}</span>
</div>
```

^^In this case, the `<input>` is the parent template, so since the `<span>` is outside of that, it doesn't work. Is my understanding correct?

Then, in the third example:

```
<ng-container *ngFor="let i of [1,2]">
  <input #ref type="text" [value]="i" />
</ng-container>

<!-- What will / should this display? -->
{{ ref.value }}
Since the template created through the structural directive is instantiated twice in this case, it'd be impossible to define sensibly what the reference should actually refer to. Since structural directives such as *ngIf are not actually treated special by the Angular compiler, there is no way of knowing if a template is ever instantiated, and if so, how often. Therefore accessing the reference variable from outside its parent template would not be well-defined behavior.
```
The reason it doesn't work is because `{{ ref.value }}` is outside of the template that `*ngFor` is creating, right? (Because if I put it inside, it picks up each value, 1 and 2).

And for this part:

```Since structural directives, such as <code>*ngIf</code> are not actually treated special by the Angular compiler...```

What does "treated special" refer to?

And finally here:
```...accessing the reference variable from outside its parent template would not be well-defined behavior. ```

By "well-defined" does that mean it's not a best practice?

Please excuse all the questions (I feel like a Vulcan! 🖖), but your guidance is REALLY helpful and I want to understand thoroughly so I can add it to the documentation.

Reviewer: > ^^This works because we access ref1 within the template created by the *ngIf (that is, we're still inside the <span>)?

Mh, not quite. In this case we have a ("hidden") `ng-template` around the `span` and the definition of the variable is outside of it. So basically we access a reference variable from a parent template (which is fine as the context is inherited). Maybe it helps to manually desuggar the syntax:

```
<input #ref1 type="text" [(ngModel)]="firstExample" />

<!-- Uh-oh, a new template! -->
<ng-template [ngIf]="true">
  <!-- … but it works anyway :-) -->
  <span>Value: {{ ref1.value }}</span>
</ng-template>
```

---

> ^^In this case, the <input> is the parent template, so since the <span> is outside of that, it doesn't work. Is my understanding correct?

Here it's more about the opposite of the above. Again, let's desuggar:

```
<ng-template [ngIf]="true">
  <!-- The reference is defined within a template -->
  <inpu #ref2 type="text" [(ngModel)]="secondExample" />
</ng-template>

<!-- … and then accessed from outside that template, that doesn't work -->
<span>Value: {{ ref2?.value }}</span>
```

---

> The reason it doesn't work is because {{ ref.value }} is outside of the template that *ngFor is creating, right? (Because if I put it inside, it picks up each value, 1 and 2).

Correct; that's exactly the same as above. Here you explained it correctly, above it seemed like there was still a misunderstanding. :-)

---

> What does "treated special" refer to?

My point here is this: the `*ngFor` example makes total logical sense, there's simply no good answer to whether it should show "1" or "2". It makes sense because ngFor will instantiate the template many times over.

ngIf, on the other hand, will at most instantiate the template once. So naively I could ask why accessing the variable from outside couldn't at least work for ngIf. Since ngIf and ngFor come built-in with Angular, users often think these are magical things the Angular compiler knows about, but in reality they are simple implementations of structural directives.

If, for example, we were to build ngIf directly into the compiler, we could probably make accessing the variable work for it: logically speaking it would make sense to allow it, but it currently simply doesn't work because the compiler doesn't actually know anything about ngIf, all it sees is templates and templates in general can be instantiated arbitrarily often.

> By "well-defined" does that mean it's not a best practice?

No, I mean that there is no way to answer the question sensibly. It's a bit of a computer science / math term. Think division by zero: it isn't an undefined operation because mathematicians like to torture students, but because any answer we assign to it would lead to contradictions.

Similarly, in the ngFor case any answer we give would be logically inconsistent (it cannot be both "1" and "2" at the same time).

---

My standard disclaimer is that I don't know anything about the internals of the Angular compiler, most of this is just my observation and general experience :-)

Author: Ah, thank you so much @Airblader! Mind if I tag you when I've got this in? Really, this is incredibly helpful! 🌈🙏🏼😊

Reviewer: Yeah, of course. My pleasure! Thanks for taking the time to improve the documentation!

The scope of a reference variable is the entire template. So, don't define the same variable name more than once in the same template as the runtime value will be unpredictable.

### Template reference variable scope

> You can refer to a template reference variable anywhere directly

Reviewer 2: ```suggestion
You can refer to a template reference variable anywhere
```
Does the word `directly` really help here?

within its surrounding template.

> Note, however, that [structural directives](guide/template-syntax#built-in-structural-directives), such as `*ngIf` and `*ngFor`, or `<ng-template>` act as a template boundary outside of which the template reference variable cannot be accessed.

Reviewer 2: I would make line 1810 an actual note, or simply write as follows:

```suggestion
[Structural directives](guide/template-syntax#built-in-structural-directives), such as `*ngIf` and `*ngFor`, or `<ng-template>` act as a template boundary. You cannot access template reference variables outside of these directives.
```

#### Accessing in a child component template

> You can access a template reference variable from an inner template because the context is inherited.

Reviewer 2: Is inherited by what? Can we use active voice here?

In the following example, changing the text in the `<input>` changes the value in the `<span>` because updates are immediately reflected through the template reference variable, `ref1`.

<code-example path="template-reference-variables/src/app/app.component.html" region="template-ref-vars-scope1" header="src/app/app.component.html"></code-example>

In this case, there is an implied `<ng-template>` around the `<span>` and the definition of the variable is outside of it.

> As this context is inherited, accessing a reference variable from the parent template works.

Reviewer 2: Rewrite to remove passive voice, if possible.

Manually desugaring the syntax illustrates this point:

```
<input #ref1 type="text" [(ngModel)]="firstExample" />

<!-- New template -->
<ng-template [ngIf]="true">
  <!-- Since the context is inherited, the value is available to the new template -->
  <span>Value: {{ ref1.value }}</span>
</ng-template>

```

However, accessing from outside the parent template doesn't work:

```html
  <input *ngIf="true" #ref2 type="text" [(ngModel)]="secondExample" />
  <span>Value: {{ ref2?.value }}</span> <!-- doesn't work -->
```

> Desugaring shows that `ref2` is outside the parent template.

Reviewer 2: I get that desugaring is a term. But is it the right term? I would almost prefer "Removing the syntactic sugar..."

```
<ng-template [ngIf]="true">
  <!-- The reference is defined within a template -->
  <input #ref2 type="text" [(ngModel)]="secondExample" />
</ng-template>

<!-- ref2 accessed from outside that template doesn't work -->
<span>Value: {{ ref2?.value }}</span>

```

Another example uses `*ngFor`. Consider the following code:

```
<ng-container *ngFor="let i of [1,2]">
  <input #ref type="text" [value]="i" />
</ng-container>

{{ ref.value }}
```

Here, `ref.value` doesn't work.
The structural directive, `*ngFor` instantiates the template twice (because `*ngFor` iterates over the two items in the array), so it is impossible to define what the reference actually refers to.

> With structural directives, such as `*ngIf`, there is no way for Angular to know if a template is ever instantiated, and if so, how often.

Reviewer 2: ```suggestion
With structural directives, such as `*ngIf`, there is no way for Angular to know if a template is ever instantiated.
```

I think this reads better.

> Therefore, accessing the reference variable from outside its parent template is not  well-defined behavior, that is, Angular isn't able to access the value and returns an error.

Reviewer 2: ```suggestion
As a result, Angular isn't able to access the value and returns an error.
```

The scope of a reference variable is the entire template. So, don't define the same variable name more than once in the same template as the runtime value will be unpredictable.

### Template reference variable scope

You can refer to a template reference variable anywhere within its surrounding template.

> [Structural directives](guide/template-syntax#built-in-structural-directives), such as `*ngIf` and `*ngFor`, or `<ng-template>` act as a template boundary. You cannot access template reference variables outside of these directives.

Reviewer 3: "outside of these directives." -> "outside of these boundaries"? or "outside of boundaries created by these directives"?

 I'm saying this since we can have `<ng-template>...</ng-template>`  (scope / boundary)  without any structural directives. What matters is the boundary created by `<ng-template>`. NgIf / NgFor happen to _need_ `<ng-template>`

> #### Accessing in a child component template

Reviewer 3: Hmm, I would probably steer away from using the "child component template" and would suggest "nested template" or "inner template".  "child component template" sounds like we might be talking about a template of a child component, so we might confuse users.

> You can access a template reference variable from an inner template because the template inherits its context from the parent template.

Reviewer 3: "because the template inherits its context" -> "because the template can access reference variables defined in the outer template"? I'm trying to avoid the "inheritance" term as we are nesting scopes here so talking about inheritance might give people "wrong" mental picture.

"from the parent template" -> "from the outer template"? Generally speaking I would probably avoid using parent / child in favour of inner / outer.
